src/common/task_manager/manager.rs

```diff
@@ -0,0 +1,214 @@
+    /// # Errors
+    /// Returns `Err` if the task manager is shutting down.
+    #[inline]
+    pub async fn spawn<F, Fu>(&self, name: TaskName, f: F) -> Result<(), SpawnError>
+    where
+        F: FnOnce(CancellationToken) -> Fu,
+        Fu: Future<Output = ()> + Send + 'static,
+    {
+        let mut tasks = self.tasks.lock().await;
+        if self.is_shutdown() {
+            return Err(SpawnError(name));
+        }
+        let node = tasks
+            .get_mut(&name)
+            .unwrap_or_else(|| unreachable!("Task {name:?} is not in the manager."));
+
+        node.spawn(f).await
+    }
+
+    /// The status of task manager, `true` for shutting down.
+    #[inline]
+    pub fn is_shutdown(&self) -> bool {
+        self.status.load(Ordering::Acquire)
+    }
+
+    /// Shutdown the task manager.
+    ///
+    /// After `shutdown` being called, no new task should be spawned via task
+    /// manager.
+    #[inline]
+    #[instrument(skip(self))]
+    pub async fn shutdown(&self) {
+        let mut queue = VecDeque::from([TaskName::Root]);
+
+        self.status.store(true, Ordering::Release);
+
+        let mut tasks = std::mem::take(&mut *self.tasks.lock().await);
+
+        while let Some(task_name) = queue.pop_front() {
+            let Some(mut task_node) = tasks.remove(&task_name) else {
+                continue;
+            };
+
+            info!("Shutdown task node: {task_name:?}");
+
+            // Notify all pending tasks to shutdown. And wait all of them to finish or quit.
+            task_node.token().cancel();
+            for result in task_node.join_all().await {
+                if let Err(e) = result {
+                    error!("Background task {task_name:?} failed with error: {e}.");
+                }
+            }
+
+            for dependency_name in task_node.dependencies() {
+                let Some(dependency_node) = tasks.get_mut(dependency_name) else {
+                    continue;
+                };
+                dependency_node.dec_predecessor_count();
+                if dependency_node.predecessor_count() == 0 {
+                    queue.push_back(*dependency_name);
+                }
+            }
+        }
+    }
+}
+
+impl Default for TaskManager {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Wait for signal `SIGTERM`, `SIGQUIT` and `SIGINT`, and shutdown the task
+/// manager.
+#[inline]
+pub fn wait_for_shutdown(
+    manager: &TaskManager,
+) -> anyhow::Result<impl Future<Output = ()> + Send + '_> {
+    use signal_hook::consts::TERM_SIGNALS;
+
+    let mut signals = Signals::new(TERM_SIGNALS)?;
+    let handle = signals.handle();
+
+    let future = async move {
+        if let Some(signal) = signals.next().await {
+            debug_assert!(
+                TERM_SIGNALS.contains(&signal),
+                "The signal hook is not to handle signal {signal}."
+            );
+            info!("Signal {signal} raised, start to shutdown.");
```

Comment:
It's a warn.

Reply:
I don't think it's a warn. A signal to shutdown is such a normal situation, such as `Ctrl+C`, `kill` and `docker stop`.